src/meta/backup_engine.cpp

```diff
@@ -184,6 +185,94 @@ void backup_engine::on_backup_reply(error_code err,
                                     gpid pid,
                                     const rpc_address &primary)
 {
+    ddebug_f("backup_id({}): receive backup response for partition {} from server {}, rpc error "
+             "{}, response error {}.",
+             _cur_backup.backup_id,
+             pid.to_string(),
+             primary.to_string(),
+             err.to_string(),
+             response.err.to_string());
+    dassert_f(response.pid == pid,
+              "backup partition id {} vs {} don't match",
+              response.pid.to_string(),
+              pid.to_string());
+    dassert_f(response.backup_id == _cur_backup.backup_id,
+              "backup id {} vs {} don't match",
+              response.backup_id,
+              _cur_backup.backup_id);
+
+    {
+        zauto_lock l(_lock);
+        // if backup of some partition failed, we would not handle response from other partitions.
+        if (is_backup_failed) {
+            return;
+        }
+    }
+
+    // if backup completed, receive ERR_OK and
+    // resp.progress=cold_backup_constant::PROGRESS_FINISHED;
+    // if backup failed, receive ERR_LOCAL_APP_FAILURE;
+    // backup not completed in other cases.
+    // see replica::on_cold_backup() for details.
+    int32_t partition = pid.get_partition_index();
+    if (err == dsn::ERR_OK && response.err == dsn::ERR_OK &&
+        response.progress == cold_backup_constant::PROGRESS_FINISHED) {
+        {
+            zauto_lock l(_lock);
+            _backup_status[partition] = backup_status::COMPLETED;
+        }
+        complete_current_backup();
+        return;
+    }
+    if (response.err == ERR_LOCAL_APP_FAILURE) {
+        derror_f("backup_id({}): backup for partition {} failed.",
+                 _cur_backup.backup_id,
+                 pid.to_string());
+        zauto_lock l(_lock);
+        is_backup_failed = true;
+        _backup_status[partition] = backup_status::FAILED;
+        return;
+    }
+
```

Comment:
It seems that you don't handle rpc error specially. If rpc errors happen, it will call function `backup_app_partition`, and let `_backup_status[pidx] = backup_status::ALIVE;`, like normal not-finished case. How about adding some comments or log here?